src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp

```diff
@@ -1686,14 +1686,140 @@ INSStaggeredHierarchyIntegrator::regridHierarchyBeginSpecialized()
     d_div_U_norm_1_pre = d_hier_cc_data_ops->L1Norm(d_Div_U_idx, wgt_cc_idx);
     d_div_U_norm_2_pre = d_hier_cc_data_ops->L2Norm(d_Div_U_idx, wgt_cc_idx);
     d_div_U_norm_oo_pre = d_hier_cc_data_ops->maxNorm(d_Div_U_idx, wgt_cc_idx);
+
+    // Free vector space.
+    if (d_U_rhs_vec)
+    {
+        d_U_rhs_vec->freeVectorComponents();
+        d_U_rhs_vec.setNull();
+    }
+    if (d_U_adv_vec)
+    {
+        d_U_adv_vec->freeVectorComponents();
+        d_U_adv_vec.setNull();
+    }
+    if (d_N_vec)
+    {
+        d_N_vec->freeVectorComponents();
+        d_N_vec.setNull();
+    }
+    if (d_P_rhs_vec)
+    {
+        d_P_rhs_vec->freeVectorComponents();
+        d_P_rhs_vec.setNull();
+    }
+
+    for (auto& nul_vec : d_nul_vecs)
+    {
+        if (nul_vec)
+        {
+            nul_vec->freeVectorComponents();
+            nul_vec.setNull();
+        }
+    }
+
+    for (auto& U_nul_vec : d_U_nul_vecs)
+    {
+        if (U_nul_vec)
+        {
+            U_nul_vec->freeVectorComponents();
+            U_nul_vec.setNull();
+        }
+    }
+
+    d_vectors_need_init = true;
+
+    // Deallocate linear solvers.
+    if (d_velocity_solver) d_velocity_solver->deallocateSolverState();
+    if (d_pressure_solver) d_pressure_solver->deallocateSolverState();
+    if (d_stokes_solver) d_stokes_solver->deallocateSolverState();
+
+    d_velocity_solver_needs_init = true;
+    d_pressure_solver_needs_init = true;
+    d_stokes_solver_needs_init = true;
+
     return;
 } // regridHierarchyBeginSpecialized
 
 void
 INSStaggeredHierarchyIntegrator::regridHierarchyEndSpecialized()
 {
-    const int wgt_cc_idx = d_hier_math_ops->getCellWeightPatchDescriptorIndex();
+    // Reset hierarchy dependent data and objects.
+    const int coarsest_hier_level = 0;
+    const int finest_hier_level = d_hierarchy->getFinestLevelNumber();
+
+    // Reset the hierarchy operations objects for the new hierarchy configuration.
+    d_hier_cc_data_ops->setPatchHierarchy(d_hierarchy);
+    d_hier_cc_data_ops->resetLevels(coarsest_hier_level, finest_hier_level);
+
+    d_hier_fc_data_ops->setPatchHierarchy(d_hierarchy);
+    d_hier_fc_data_ops->resetLevels(coarsest_hier_level, finest_hier_level);
+
+    d_hier_sc_data_ops->setPatchHierarchy(d_hierarchy);
+    d_hier_sc_data_ops->resetLevels(coarsest_hier_level, finest_hier_level);
+
+    if (d_manage_hier_math_ops)
+    {
+        d_hier_math_ops->setPatchHierarchy(d_hierarchy);
+        d_hier_math_ops->resetLevels(coarsest_hier_level, finest_hier_level);
+    }
```

Comment:
This should not be necessary - the base class already executes this check. We should not need to mess around with objects managed by the base class.